### PR TITLE
Backport service discovery related changes on 3.0 to master

### PR DESCRIPTION
**Pull request by chickenlj**

### CODE REVIEW

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/transport/AbstractEndpoint.java [hunk 1]
```diff
@@ -51,7 +51,7 @@ public AbstractEndpoint(URL url, ChannelHandler handler) {
     }
 
     protected static Codec2 getChannelCodec(URL url) {
-        String codecName = url.getParameter(Constants.CODEC_KEY, "telnet");
+        String codecName = url.getProtocol(); // codec extension name must stay the same with protocol name
```
Comment (htynkn) [inline dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/transport/AbstractEndpoint.java:54]: what's the purpose for this change?
does this a break change?

Comment (htynkn) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/transport/AbstractEndpoint.java:54]: @chickenlj

Comment (ririliu) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/transport/AbstractEndpoint.java:54]: same question
